### DISCUSSION TITLE
FFMPEG Video Filters and 'Frame Rate Override' - v0.12.0 Alpha 134

### DISCUSSION THREAD

**Discussion by Notheisz57** [Xaymar/obs-stream-effects#956] [Q&A] [answered]
I am using custom FFMPEG video filters that modify the video output frame rate, and testing/reviewing the result via local recording.
There is no "do not override" option for 'Framerate Override', so I set this option to what my filter graph should be outputting, and it worked with surprising success!
That said, the video stream in the resulting `MKV` file describes itself as the original OBS frame rate.  I assume this is actually expected behavior, and is only really noticeable when writing to a medium that utilizes a file container (ex. file storage vs internet streams)

Doing an actual frame count externally via FFMPEG and some math, the result appears to be what I actually set for 'Framerate Override'.
For example, running:  `C:\ffmpeg\bin\ffmpeg.exe -i ".\2022-10-11 08-52-56.mkv" -map 0:v:0 -c copy -f null -`
Initially describes video stream: `Stream #0:0: Video: h264 (High), yuv420p(tv, bt709, progressive), 1920x1080 [SAR 1:1 DAR 16:9], 120 fps, 120 tbr, 1k tbn`
Then returns with: `frame= 7342 fps=0.0 q=-1.0 Lsize=N/A time=00:02:02.33 bitrate=N/A speed=1.48e+03x`
Maths: `7342 / 122 == 60.018`

I assume this is due to some sort of limitations in OBS plugins, specifically something like OBS must handle file creation (stream headers/metadata), and is probably what was described by ["... quite literally breaking several specifications..." on the release page](https://github.com/Xaymar/obs-StreamFX/releases#:~:text=Note%20that%20this%20is%20quite%20literally%20breaking%20several%20specifications%2C%20so%20use%20at%20your%20own%20risk).

Either way, I was still initially confused when VLC (_not a great test, I know_) reported the file as the original OBS framerate.
Is it maybe possible to add one of those little `(i)` tooltips for this option, like the ones you see on OBS built-in NVENC encoder options for tuning and what-not?
It could say something like: `Caution: FPS reported by file container (media players) may not reflect actual frame count of the output video stream`

This is a fantastic option/feature - Please do not remove it if someone gets noisy about it later!

**Comment by Notheisz57** [comment]
Sorry I do not think is actually working as I had thought.  Maybe I misunderstood the intended use of this input option?
Upon looking through the logs, I see warnings regarding the custom FFMPEG parameters I specify.

```
15:09:01.084: [StreamFX] Option '-vf' is malformed, must contain a '='.
15:09:01.084: [StreamFX] Option 'zscale=1920:1080:filter=spline36' is malformed, must start with a '-'.
```
I've tried a variety of different escapings and such with no success.  I referenced the code around here to get some guidance:
https://github.com/Xaymar/obs-StreamFX/blob/6eb566369120d7d64b0642204c33b14a42977f56/source/encoders/encoder-ffmpeg.cpp#L930

As far as running FFMPEG on the command line, splitting parameter KV sets on spaces does not align with my understanding of FFMPEG parameter usage.
Spaces are the commonly used separator between the key and the value, and attempting to use an equals sign results in an error from FFMPEG.

Here is an example of a filter graph parameter I was trying to use: `-vf zscale=1920:1080:filter=spline36`
Use it in a sentence: `ffmpeg.exe -i ".\2022-10-11 11-58-47.mkv" -vf zscale=1920:1080:filter=spline36 -vcodec hevc_nvenc -preset lossless '.\2022-10-11 11-58-47.test.mkv'`

If this should be doable, but the argument parsing needs a change, I can make a pull request, but I assume I misunderstand how this option can actually be used.
Do you mind clarifying if this should be possible?

**Comment by Xaymar** [comment]
StreamFX only provides FFmpeg Encoders, not all of FFmpeg. This is already clearly explained on the wiki.

**Reply by Xaymar** [reply]
This means that you can use the options that FFmpeg tells you are available when you run `ffmpeg -help encoder=ENCODERNAME`, and some core video options. If you want to use FFmpeg directly, you need to use Custom Output.

**Comment by Xaymar** [comment, marked as answer]
> I am using custom FFMPEG video filters that modify the video output frame rate, and testing/reviewing the result via local recording.

As explained in [the wiki](https://github.com/Xaymar/obs-StreamFX/wiki), StreamFX only integrates Encoders. Therefore you can't expect it to do anything that isn't encoding - it's not FFmpeg, just the encoders in FFmpeg.

> There is no "do not override" option for 'Framerate Override', so I set this option to what my filter graph should be outputting, and it worked with surprising success!

As explained in [the wiki](https://github.com/Xaymar/obs-StreamFX/wiki/Encoder-FFmpeg#framerate-override), the default setting will not skip any frames. 

> That said, the video stream in the resulting MKV file describes itself as the original OBS frame rate. I assume this is actually expected behavior, and is only really noticeable when writing to a medium that utilizes a file container (ex. file storage vs internet streams)
> I assume this is due to some sort of limitations in OBS plugins, specifically something like OBS must handle file creation (stream headers/metadata), and is probably what was described by ["... quite literally breaking several specifications..." on the release page](https://github.com/Xaymar/obs-StreamFX/releases#:~:text=Note%20that%20this%20is%20quite%20literally%20breaking%20several%20specifications%2C%20so%20use%20at%20your%20own%20risk).

As explained in [the wiki](https://github.com/Xaymar/obs-StreamFX/wiki/Encoder-FFmpeg#framerate-override) and the release notes, the behavior is expected. 

> Is it maybe possible to add one of those little (i) tooltips for this option, like the ones you see on OBS built-in NVENC encoder options for tuning and what-not?

The use of the "long description" feature in OBS Studio was removed several versions ago, due to there being no easy way to update these in all languages. The wiki is now used for all documentation purposes, as StreamFX is designed and meant for power users.

> As far as running FFMPEG on the command line, splitting parameter KV sets on spaces does not align with my understanding of FFMPEG parameter usage. Spaces are the commonly used separator between the key and the value, and attempting to use an equals sign results in an error from FFMPEG.

As you can probably already imagine, StreamFX is not FFmpeg, and does not aim to be FFmpeg. [The wiki has an exact description of how to use the `Custom Settings` field.](https://github.com/Xaymar/obs-StreamFX/wiki/Encoder-FFmpeg#custom-settings)

> Do you mind clarifying if this should be possible?

It is not possible to use features that don't exist in StreamFX. Surprisingly, you're not the first one to do this, and I don't understand how this mistake keeps being repeated. StreamFX has never advertised, documented, or instructed people to do this, yet it is still something people do. :(

**Reply by Notheisz57** [reply]
No worries!   I appreciate the detailed response and wiki links.
I had looked on the wiki, but at the time I didn't see a description for the FFMPEG arguments.  I likely ended up on the wrong page; I remember reading descriptions for the NVENC options but not seeing anything for this custom settings input.

When I saw the warning logs I began thinking as you said; I know it is uncommon ship anything with an entire build of FFMPEG.  Integrating the libs is more common.  Sorry for adding to the statistic 😭

For what it's worth, this was not an original goal for me.  I was more-so just playing with the tool to assess what was most feasible for my goals.
The tool provides all my initially desired features, and is an incredible asset and improvement on OBS as it is 🙂


When I started using the plugin, I faced some strange issues that made me think "race condition", but this is barely more than a guess and provides no real help to you on correcting the issue.
I am not limited by color format or system resources, so I switched from the NV12 color format to I420 after seeing the similar issue is currently open. (Can provide more details there if you want)

The original issue was simply the description regarding frame rate overrides.
When I faced the other issues I got a bit sidetracked and stole your time when I shouldn't have, I apologize.

**Reply by Xaymar** [reply]
Obs itself ships with the entirety of FFmpeg, but that was not the goal of the encoder integration. You can already use FFmpeg with the custom output mode
